### DISCUSSION TITLE
[Feature]: Support more storage backend

### DISCUSSION THREAD

**Discussion by gromain** [immich-app/immich#1683] [Feature Request]
### Feature detail

Right now, only a local folder is supported natively. Even though more can be supported through fuse mounted filesystems, but this method has drawbacks (performance hit mainly), so it would be nice to have support for more back-ends eventually.

S3 is an obvious choice that comes to mind and that would support a lot of providers, but there could be others (like WebDAV perhaps).

### Platform

Server

**Comment by Nonobis** [comment]
Minio would be great !

**Comment by gromain** [comment]
Minio is compatible with S3!

**Comment by Nonobis** [comment]
@gromain : i know :)

**Comment by palitu** [comment]
Would this be local storage for thumbnails, and remote storage for files?  That would maintain speed for scrolling.

Then you would need to pass a Uri to the app to retrieve the full file for viewing?

**Comment by akashkj** [comment]
If it allows to have a backup also for the media, that would be awesome.
Like 2 storage backends - s3 and G-drive. If one of them is not reachable, switch to other. This could be defined in a priority list.

**Comment by gromain** [comment]
In my mind it could have been for all objects, but @palitu you're right that thumbnails could still be local.
And yes, @akashkj, that's how I felt about it. Having two or more storage back-ends possible, that can either work as RAID0 or RAID1 (not exactly RAID, but either mirror or concatenate the available storage).

**Reply by tarasis** [reply]
I would definitely love to see this (I guess I could use a symbolic link for the upload folder). I have a Hetzner Storage Box with 5TB of space, and would love to use it to store the photos, and leave the thumbnails local to the host running immich. (free Oracle Cloud instance). 

That way I have some additionally safety as they take snapshots of the data.

**Comment by zackpollard** [comment]
In my mind, supporting a well defined platform such as S3-like APIs is a good idea and is something we can look into after the first major release is completed. As for the other idea of having multiple backends that we switch between, in my mind that is the job of the storage layer, not something that immich should natively support. We'll leave this here and probably get around to it at some point in the future!

**Comment by palitu** [comment]
> 

Really great to hear your desire for an S3 type backend.  FYI, i use the local filesystem for most of immich, then a mounted NFS share for the main photo storage.  My server doesnt have the local storage for the bulk data (especially with all the video's), but thumbs is more than appropriate for local.

I do think that there are some delays with video playing as a result of this, but i assume in the future it will be improved with this S3 feature.

ta

**Comment by Cellivar** [comment]
I just started writing this up as a new feature request before discovering this one. I would love to be able to use S3 storage _for all blob objects_ and make _a raw disk volume optional_. Big +1 to this feature request!

At home I run a Nomad cluster, like a Kubernetes cluster but with less yaml involved. We use CSI-managed iSCSI storage volumes that are mounted directly into running docker images (or other tasks). The preference here is a single writer job and _maybe_ multiple read-only mounts for other volumes. This means that _sharing a directory is hard_ while _communicating over the network is easy_. Right now the docker-compose files seem to suggest mounting one volume to multiple hosts that all have write access which is a little scary (and hard to do with iSCSI). The performance hit of switching to NFS is very visible on our network.

There's an [official TS SDK](https://github.com/minio/minio-js) for all the bells and whistles MinIO provides, but I'd strongly suggest looking at just [using the S3 SDK which would allow folks to use any S3-compatible storage backend](https://northflank.com/guides/connect-nodejs-to-minio-with-tls-using-aws-s3).

This is of course asking a lot, adding a new storage backend like S3 is completely different than writing to disk 😄 I'm perfectly happy to be patient for this sort of feature.

**Comment by athornfam2** [comment]
Would like to add that I am interested in S3 storage access as well to offload.

**Comment by Zen3515** [comment]
I try mounting S3 with rclone and found that immich uses softlink which is not allowed in S3.  

I guess this is one of the things to consider when implement this in the future.

```shell
Error: ENOSYS: function not implemented, link 'upload/04f8f839-9ee4-47fc-89b1-fb3883cf3008/original/WEB/33809e6c-1f1f-4bec-a8cc-7a0367493e8d.png' -> 'upload/04f8f839-9ee4-47fc-89b1-fb3883cf3008/2022/2022-12-11/hahaha.png'
```

**Reply by nperez0111** [reply]
I ran into this same issue while trying to use a webdav mount via `davfs2`

**Reply by NiklasRosenstein** [reply]
I've run into the same issue, sadly. If it didn't try to use soft links, I think Immich could already run pretty well using S3 fuse mounted storage!

**Reply by jcollie** [reply]
Had similar issues trying to run Immich in Azure/Azure Kubernetes. Azure does not really have a good ReadWriteMany solution for sharing volumes between pods since it's using CIFS under the hood.

**Reply by carballude** [reply]
Same thing using SMB mounts:
`ERROR [StorageTemplateService] Error: ENOTSUP: operation not supported on socket, link '<<source>>' -> '<<destination>>'`

**Reply by NiklasRosenstein** [reply]
It works well with the yandex csi-s3 driver using geesefs.

**Comment by nicolopozzato** [comment]
Hi, i setup immich on a arm vps on oracle, with ocamel fuse google drive, but I have the same error:

immich_server | [Nest] 1  - 03/01/2023, 12:47:57 AM ERROR [StorageService] Error: ENOSYS: function not implemented, link 'upload/85ff3fb7/original/WEB/408963f5-45f7-40e8-960c-79f36116e086.jpeg' -> 'upload/85ff3fb7/2023/2023-02-25/originalName.jpeg'

So in the folder 2023-02-25 the image is not present but it's only in the original/WEB/....

**Comment by cycneuramus** [comment]
If I'm reading the source code correctly, the reason Immich currently doesn't support `rclone` mounted backends is simply because: 

+ its [way of moving files ](https://github.com/immich-app/immich/blob/main/server/src/infra/repositories/filesystem.provider.ts#L8) is by relying on the [mv library](https://github.com/andrewrk/node-mv), which:
    + works by calling [fs.link](https://github.com/andrewrk/node-mv/blob/master/index.js#L39) (at least on `clobber: false`, which Immich is currently using), which: 
        + creates a hard link of the source file at the destination path, at which:
            + `mv` will unlink the source file, thus having "moved" the file from source to destination.

Now, the problem is, `rclone` doesn't support hard links. Hence the move operation fails halfway with the `ENOSYS: function not implemented, link ...` error. 

I don't have a thorough understanding of the way Immich is supposed to work here, but if the point is to move the files and nothing more, it seems to me we could achieve full `rclone` compatibility—and with it support for countless storage backends—simply by reworking the move command (say, by using `fs.rename(sourcePath, desPath);` or something).

I've created an issue about this [here](https://github.com/immich-app/immich/issues/2831).

EDIT: PR [here](https://github.com/immich-app/immich/pull/2832)

EDIT2: PR is now [merged](https://github.com/immich-app/immich/pull/2832), so all `rclone` supported remotes should work with Immich probably in the next release.

**Reply by kajdo** [reply]
great!!!

**Reply by pinpox** [reply]
Is rclone the way to go for using immich with s3 currently? Are there plans to support s3 natively?

**Comment by Koveras-hmm** [comment]
webdav plz!

**Reply by cycneuramus** [reply]
Already supported, at least indirectly, via [rclone](https://rclone.org/webdav).

**Comment by alkmim** [comment]
FYI: I did some tests using rclone and I reported my findings (without a lot of data though) here: https://github.com/immich-app/immich/discussions/3814. 

In a nutshell, it did not work reliably enough for me. 

Maybe we should not consider this feature supported (indirectly via rclone)

**Comment by david-polak** [comment]
I just installed Immich on k8s and connected it to an s3 compatible storage, all hosted on Vultr. It's bit of a kludge setup and took me four hours to debug.

Immich pods - mounted nfs dir -> NFS server pod - mounted s3 dir using s3fs -> s3

The results, well it works. Kinda, if parallelism is reduced to 1 it almost handles uploads of regular photos. There are missed thumbnails here and there and corresponding filesystem errors. Videos are mostly broken, haven't had time to see what exactly is the issue with those apart from ffmpeg errors.

Needs more investigation, these are my initial impressions so far.

**Reply by NiklasRosenstein** [reply]
I'm running things on a CIFS mount as well and it's pretty smooth sailing. Haven't reached the level of brave required to nest an S3 mount inside it, though.

**Comment by mariusm0e** [comment]
I am currently using rclone to mount google drive with seemingly no errors using vfs-cache-mode write. I only mounted the library and upload folder of the docker containers from the rclone folder, having the thumbnails mounted from the server locally.

`rclone mount drive: /mnt/drive --vfs-cache-mode writes -vv`

**Reply by StanfordMATT** [reply]
Can you share a Docker Compose? Interested in using Google Drive as a persistence layer

**Reply by mariusm0e** [reply]
I'm using ansible instead of docker-compose, but its not really anything special. Anywhere where normally the "upload" folder is mounted into the container, I instead mounted the individual folders inside. 

For the microservices container, it looks like this (again not docker-compose syntax)
```
    mounts:
      - source: "{{ immich.upload_path }}"
        target: /usr/src/app/upload/upload
        type: bind
      - source: "{{ immich.library_path }}"
        target: /usr/src/app/upload/library
        type: bind
      - source: "{{ immich.thumbs_path }}"
        target: /usr/src/app/upload/thumbs
        type: bind
      - source: "{{ immich.media_path }}"
        target: /mnt/media
        type: bind
        read_only: true
```
        

To make sure that rclone mounts before the containers mount, this is the systems service that was set up:

```
[Unit]
Description=mount rclone drive
After=network.target
# run before docker
Before=docker.service

[Service]
Type=simple
Restart=on-failure
ExecStart=/usr/bin/rclone mount google-drive: /mnt/google-drive --vfs-cache-mode writes --vfs-write-back 5m --vfs-cache-min-free-space 5G -vv
TimeoutStartSec=20s

[Install]
WantedBy=multi-user.target
```

In my example, the folder /mnt/google-drive with the folders you want to save on the rclone mount must be created first. I now only save the "library" folder on the rclone mount, and have the rest on the server itself. be aware, the "thumbs" folder will grow quite a bit, as it has all the images you normally see when using immich, the "library" folder only stores the original images.

**Reply by stv0g** [reply]
RClone provides a Docker Volume Plug-in which can make the integration a bit nicer: https://rclone.org/docker/

But still this is just a workaround in my opinion. If supported by the storage backend, up and downloads should target the storage backend directly rather than beeing proxied by Immich or some Fuse file system as it will never scale as good as a direct access.

**Comment by janbuchar** [comment]
I was thinking about contributing s3 support to immich. I found that

1. the uploads are stored directly to the library by multer - I could probably use multer-s3 to take care of this
2. exiftool needs a local copy of the file, but it shouldn't be a problem to download a temporary copy just for metadata extraction
3. most other filesystem-related functionality is handled by `IStorageRepository`, it should be enough to make an alternative implementation of that
4. files are served using sendFile from express, and  it may be tricky to adjust serveFile to support direct streaming from s3

Am I missing some other roadblocks? Would this level of support for s3 even be welcome?

**Reply by pinpox** [reply]
Please do, I'm eagerly awaiting this feature. If you need help testing let me know!

**Reply by janbuchar** [reply]
I'd also love to have it, but I really need some feedback from the team before. I'd hate to spend time on something that wouldn't get merged in the end :slightly_smiling_face:

**Reply by janbuchar** [reply]
It seems that @uhthomas plans to work on this as well - see https://github.com/immich-app/immich/issues/4445

**Reply by david-polak** [reply]
Hmm I threw some hours at it couple of months back... https://github.com/immich-app/immich/compare/main...david-polak:immich:s3 with no avail, I got basic stuff working but not everything was configured to work with streams or the file access layer, the actual refactor needed to properly support s3 is significant and should be undertaken under the supervision and approval of the project leaders.

**Reply by stv0g** [reply]
> files are served using sendFile from express, and it may be tricky to adjust serveFile to support direct streaming from s3

I think it would be more efficient to let the client directly access the object store..

**Reply by systemmonkey42** [reply]
> I think it would be more efficient to let the client directly access the object store..

Does this mean if I share a library, I'm sharing the s3 api keys?
I'm happier trading that efficiency for the security of having the immich server securely proxy the s3 requests.

**Reply by stv0g** [reply]
> Does this mean if I share a library, I'm sharing the s3 api keys?

No, S3 has a feature called Presigned URLs, which can be used to generate short lived URLs to S3 objects without exposing the S3 credentials. That would even allow for direct client to S3 uploads.

**Reply by pinpox** [reply]
> No, S3 has a feature called Presigned URLs, which can be used to generate short lived URLs to S3 objects without exposing the S3 credentials. That would even allow for direct client to S3 uploads.

This is how most products using S3 do it, e.g. ente photos works like this.

**Reply by PatrickHuetter** [reply]
Presigned URLs is the way to go.

**Reply by uhthomas** [reply]
Presigned URLs have quite a few caveats, especially at home:
- Clients may have a much harder time caching content.
- The S3 host may not be publicly accessible.
- URLs are not friendly, and if visited directly in a browser then it won't be possible to refresh when the URL expires.
- More?...

Redirecting to a different location is not always desirable and often actually harmful. The server has much less control over the how the content is served and it can get quite messy. It also means the behaviour of how assets are served will differ hugely between local assets vs s3 vs any other storage medium, which adds complexity. IMO I would just avoid presigned URLs entirely.

I personally also quite liked the idea, until I implemented it and used it. I then removed it entirely because it has a lot of drawbacks.

**Reply by PatrickHuetter** [reply]
> Clients may have a much harder time caching content.

Could you explain why? I don't get this point since in my opinion it much better and easier to cache. Even you could cluster your pictures and have direct I/O via different IP routes over the internet to different storage servers. Also in home scenarios if someone decides to use s3 as backend he could decide wether to route traffic to a different device (e.g. NAS with s3 service while Immich is running on a different homeserver).

> The S3 host may not be publicly accessible.

If this is the case, your immich instance wouldn't be accessible too? Otherwise you have reverse tunnels like tailscale, cloudflared etc.? If this is the case you could just don't use the s3 backend or am i wrong? Since it's just a decision which storage layer you choose or isn't it?

> URLs are not friendly, and if visited directly in a browser then it won't be possible to refresh when the URL expires.

What do you mean with "urls are not friendly"? And what's bad about not beeing able to refresh after an presigned url expires? It just the way presigned URLs work. Knowing it from my own software development and also a bunch of other open source projects where this works absolutely fine. 

I can only speak for our spring boot environments where s3 integration with presigned urls works perfect without any issues. We will never look back until we find something better for storing blobs/files :)

**Comment by systemmonkey42** [comment]
+! for this to support structured storage for original media while keeping thumbnails and cache data on a local filesystem.

This would also benefit from some method to take a database dump and upload it to the same structured storage for future DR scenarios.

Currently I use snapshots for the database, but that not going to help in the event of a fire, even though all the original media is currently backed up to an S3 bucket.

**Comment by sorbanbela** [comment]
rClone support is great, but I think that native S3 support would be really nice. Serving thumbnails and images with presigned URLs would improve overall performance and scalability.

**Comment by mrvanes** [comment]
I just deployed a test instance of Immich with an S3 backed storage via https://github.com/s3ql/s3ql
Seems to fare well at first sight. I know it's not native S3, but at least I now have an elastic storage backend that will only cost what I use.

**Comment by dubrowin** [comment]
I worked this week to bring up an Immich server backed by S3. I'm using an AWS built and supported tool. Documentation on my setup is here: https://github.com/dubrowin/Immich-backed-by-S3/

**Comment by Cohinem** [comment]
I currently mount WebDAV with rclone. It works fine most of the time. but there are some problems. native support would surely be great.

**Comment by dubrowin** [comment]
In general, I found the costs of using S3 as the back-end too expensive. It was much cheaper to expand the EBS volume I'm using to a larger volume size (in GP3).

**Reply by PatrickHuetter** [reply]
I depends where you order/buy your s3 storage or self host it using minio or different s3 storage server. https://www.idrive.com/s3-storage-e2/ for example is one of the cheaper ones. But also scaling storage is much simpler using s3 since you can have different zones, cdn and divergent storage servers that provide assets via better routes to the client (e.g. using private networks, cloudflare etc.)

**Reply by AngelaDMerkel** [reply]
I have tried this now too and costs get too high if the user is consistently accessing older images or uploading and transcoding a lot of video. It's much more inexpensive to use Google Photos if you're comfortable leaving your personal data on cloud system anyways.

**Reply by FracKenA** [reply]
There is a big difference between Google Photos Cloud and AWS, Azure, GCP, etc. Cloud. With Google Photos, google controls everything, from upload to deletion, it is all google. With one of the block storage solutions, the org owner controls it. There is no one scraping it, using it, deleting it, editing it, reducing the resolution.

**Reply by CrushedAsian255** [reply]
I use Backblaze and it’s pretty good value, $6/TB

**Reply by archerallstars** [reply]
I use Storj, $4/TB.

**Reply by dillfrescott** [reply]
Storj is not just $4 a TB but also $7 a TB of ingress

**Reply by archerallstars** [reply]
@dillfrescott Did you mean egress (download)? Since ingress (upload) has always been free on Storj. Moreover, every S3 storage costs you for egress, one way or another. For example, if you store 1 TB on [Backblaze](https://www.backblaze.com/cloud-storage/pricing), you'll have free egress until 3 TB, and another $10 a month (total of $16) for 4 TB egress.

**Reply by dillfrescott** [reply]
Sorry yeah, i couldnt think of the right word!

**Reply by leonidasv** [reply]
Be aware that EBS GP3 has 99.8%-99.9% durability, while S3 has 99.999999999%.

**Reply by mnpenner** [reply]
Backblaze has an alliance with Cloudflare, so you can get free egress if you put B2 behind a Cloudflare domain.

**Comment by surapuramakhil** [comment]
@gromain is this done / activity working on this?

**Reply by gromain** [reply]
I'm not the developer, just a user requesting a feature.

And I don't have the time to work on this unfortunately.

**Comment by justin13888** [comment]
Just adding my two cents. I've been playing around building an image/video (assets in general really) management software similar in nature to Immich but currently, specifically for temporary sharing. It purely uses S3 for storing both original copies and generated assets (e.g. thumbnails, transcoded versions).

I'll spend more time working on it again later but one thing I'd like to point out is the architecture optimized for S3 versus some filesystem-level or block-level storage looks quite different. From a developer's point of view, for example, uploading images for S3 over others could look different although sure you could just implement a way that is the least common denominator.

I'd suggest before any actual implementation work, we should gauge what type of hardware and scale people are actually using Immich for. Personally I love many things that its done for me so far but there's a lot of "cool" or high impact features will still require significant rewrites. (I'm also not a fan of nestjs backend but I digress.)

**Reply by luanjot** [reply]
I would be happy to help on this effort as I was looking for this kind of solution. Has this been discussed / decided by the community that native S3 support is something they want / don't want to do? 

I cannot see any kind of actual discussion on here on whether the feature is needed / wanted or not.

**Comment by CrushedAsian255** [comment]
Thinking of how this could actually be implemented, should only the originals be uploaded to S3 or should the thumbnails also be stored on S3?

**Reply by pinpox** [reply]
Best would be both, as one of the use-cases for this is to be able to host the application on a low-cost server which would have little storage itself

**Comment by luanjot** [comment]
The use case I have in mind is of a person that wants to have minimal
infrastructure for their photos optimising the cost. In this scenario, the
need to have the thumbnails locally would be counterproductive as you would
require either a large disk (not cost effective) or a disk you can expand
often (not minimal infrastructure).

I guess what I am asking is whether this case should be supported or not. I
think that there are lots of people that would be happy with this if set up
efficiently because it removes the need to pay for a fixed amount of
storage and the operations with thumbnails would be very few.

But maybe doing something like this moves people away from self hosting and
this is not the point of the project.

**Reply by zatricky** [reply]
The definition of "self hosting" is not so narrow. Hosting on a Raspberry Pi at home, hosting on a VPS with a cloud provider, and hosting on a physical rented server in a datacentre all qualify as self-hosted.

Additionally, block storage systems are often available for on-prem systems like Unraid or TrueNAS, typically using Minio, which is S3-compatible.

**Reply by sgofferj** [reply]
One can self-host S3-compatible storage, e.g. Minio. In fact, object storage has many advantages over block storage because many things can be handled transparently by the storage system, such as versioning, object life-cycle management, redundancy, ...
I'm self-hosting Minio and some of my use cases include:
- Storage of surveillance video with life-cycle management (automatic moving of objects from SSD to HDD after a while and automatic expiry/deletion after another while)
- Storage backend for Nextcloud (object versioning handled by Minio vs. Nextcloud native)
- Automatic replication/backups

**Reply by AngelaDMerkel** [reply]
@sgofferj This is somewhat off topic, but how can you use Minio as a storage backend for Nextcloud? Secondly, wouldn't it be somewhat redundant if using a filesystem with versioning?

**Reply by sgofferj** [reply]
https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/primary_storage.html

**Reply by sgofferj** [reply]
As far as I understand, when using S3 as backend for NC, the versioning is done through S3 natively.

**Comment by ghost** [comment]


**Reply by NiklasRosenstein** [reply]
I must say I'm running Immich quite successfully using https://github.com/yandex-cloud/geesefs (actually via https://github.com/yandex-cloud/k8s-csi-s3). It may not be perfect from an architecture point of view, but it works really well and I haven't really run into issues (aside from some woes when updating the configuration of the PVC/PV but that is more of a Kubernetes/CSI issue). You really want a good latency and bandwidth between your Immich server and the object storage, though. I'm currently going with Hetzner Cloud VPS (Helsinki) and Object Storage (Frankfurt). Despite their distance, the avg. network speed is ~450 MB/s (though I'm migrating them closer to each other soon).

**Reply by PatrickHuetter** [reply]
> I must say I'm running Immich quite successfully using https://github.com/yandex-cloud/geesefs (actually via https://github.com/yandex-cloud/k8s-csi-s3). It may not be perfect from an architecture point of view, but it works really well and I haven't really run into issues (aside from some woes when updating the configuration of the PVC/PV but that is more of a Kubernetes/CSI issue). You really want a good latency and bandwidth between your Immich server and the object storage, though. I'm currently going with Hetzner Cloud VPS (Helsinki) and Object Storage (Frankfurt). Despite their distance, the avg. network speed is ~450 MB/s (though I'm migrating them closer to each other soon).

Thanks for sharing your setup! This is exactly what i'm going to do. Good to hear this is working for you :)

**Reply by ghost** [reply]
Thanks for your reply. However, I would always favour a real implementation without relying on FUSE.

**Reply by bo0tzz** [reply]
We definitely want S3 support at some point, but it's a huge undertaking that won't happen anytime soon, as priorities are currently elsewhere. A FUSE setup like mentioned above does sort of work, but we don't recommend or support it.

**Reply by ghost** [reply]
> We definitely want S3 support at some point, but it's a huge undertaking that won't happen anytime soon, as priorities are currently elsewhere. A FUSE setup like mentioned above does sort of work, but we don't recommend or support it.

Thanks for your honest reply! I'll keep using iCloud Photos until then. Keep up the good work, Immich is amazing.

**Reply by vleeuwenmenno** [reply]
fyi, I've tried the geesefs method and it is very much not reliable enough for anything serious. I tried it on a Ampere ARM64 server from Hetzner in Helsinki with also the Object Storage in the same location.

I've tried it with Jellyfin and Immich and it often disconnected or files just disappeared out of nowhere. Reconnecting sometimes fails with weird messages that disappear after retrying. So it looks like stability varies per user since I did read some people using this but for me it'ss not.

**Reply by lukaslangrock** [reply]
Have you taken a look at JuiceFS yet? I use it for Szurubooru and Seafile and it works good, not great for Seafile as it likes to read small files squentially and also objects on Hetzner's fsn1 are currently really slow.

But for services that simply read and write larger files like Immich, this could be a great fit. It also has local caching and can be really fast, but it uses it's own filesystem ontop of s3 so that could be a dealbreaker for some.

**Reply by vleeuwenmenno** [reply]
@lukaslangrock Thanks JuiceFS was a great suggestion. I am now running Immich, Hoarder, Gitea, golink and Jellyfin in docker with data stored on a JuiceFS mount using Hetzner Object Storage. It's only running for a little bit now so we'll have to see how it holds up but so far I am very happy, it performs much better than I expected.

**Reply by Nonobis** [reply]
@vleeuwenmenno : I am interested ... how did you migrate existing data ? Move file to s3 via new folder mapped with JuiceFS and change source folder in immich or did you make a new instance of immich ?

**Reply by vleeuwenmenno** [reply]
> @vleeuwenmenno : I am interested ... how did you migrate existing data ? Move file to s3 via new folder mapped with JuiceFS and change source folder in immich or did you make a new instance of immich ?

I run Immich from the docker containers. So I stopped all containers, adjusted the volume paths, used `rsync -avR SRC DST` to copy the files to the object storage path which I had mounted using JuiceFS. And then I simply turned everything back on and it worked like it did before.

 Although it works, small update... I just had a glitch JuiceFS was mounted but no files could be read and I had to umount and remount it and restart my containers. So at the moment I am not 100% sure if this is the way to go. It makes sense that this isn't officially supported.

**Reply by vleeuwenmenno** [reply]
> > @vleeuwenmenno : I am interested ... how did you migrate existing data ? Move file to s3 via new folder mapped with JuiceFS and change source folder in immich or did you make a new instance of immich ?
> 
> I run Immich from the docker containers. So I stopped all containers, adjusted the volume paths, used `rsync -avR SRC DST` to copy the files to the object storage path which I had mounted using JuiceFS. And then I simply turned everything back on and it worked like it did before.
>
> Although it works, small update... I just had a glitch JuiceFS was mounted but no files could be read and I had to umount and remount it and restart my containers. So at the moment I am not 100% sure if this is the way to go. It makes sense that this isn't officially supported.

Last update... After running it for a weekend the SQLite database file often locked up when lots of IO operations were happening causing JuiceFS to hang and never recover, however switching backend to Redis fixed that, and it's been running smooth for almost 3 days now.

**Reply by bjo81** [reply]
Not directly related to immich, but I'm using juicefs with S3 for over a month now for Nextcloud with Valkey and it works just fine. I've tried also PostgreSQL for metadata, and it was much slower.

**Reply by vleeuwenmenno** [reply]
FYI, You're aware Nextcloud natively supports S3 as a storage option?

**Reply by NiklasRosenstein** [reply]
@vleeuwenmenno I would recommend against putting the SQLite database on the JuiceFS volume. If you still want it in S3, I suggest you use [Litestream](https://litestream.io/).

**Reply by bjo81** [reply]
Yes, but as I tried it with S3 as primary storage it had some issues:
* writes failed and NC rewrote the objects and the S3 primary storage has no garbage collector to remove them
* some writes failed, but NC thought the files existed and didn't clean up the metadata - it just kept complaing that it doesn't find the object
* something like "files --scan --all" does not work
* at least having the data within a mount on the server makes it easier to backup the data

**Reply by vleeuwenmenno** [reply]
> @vleeuwenmenno I would recommend against putting the SQLite database on the JuiceFS volume. If you still want it in S3, I suggest you use [Litestream](https://litestream.io/).

@NiklasRosenstein I am not running JuiceFS with SQLite, I am using Redis. I tried SQLite at first but this was not very optimal. With Redis however, it runs smooth as butter.

**Reply by Luncheon3462** [reply]
I need this too please

**Reply by vleeuwenmenno** [reply]
> I need this too please

It's been 3 weeks now... still running this with JuiceFS and Redis as it's backend.
LTDR: JuiceFS with Redis allows you to mount a S3 bucket as a filesystem that supports enough to make most things work fine as if it's any standard disk. It simply creates chunk files on the S3 buckets of any desired size and additionally encrypts them so your data is secured. (Not a paid advertisement btw, just a happy user!)

**Reply by NiklasRosenstein** [reply]
Ah I see, I misunderstood you. I thought you were using Immich with Sqlite and stored it on JuiceFS. I also misremembered that Immich doesn't actually support Sqlite!

**Comment by diraneyya** [comment]
This has been a fascinating and comprehensive discussion over the past two years. Thank you to everyone who has shared their experiences with various workarounds and architectural insights.

I'd like to add a perspective from the self-hosting consultant side that reinforces why native S3 support would be transformative for Immich adoption. The key architectural benefit is **decoupling compute from storage**, which offers significant advantages for both consultants and end users.

**The core use case**: Most people want Immich to help clear space on their phones by backing up photos/videos efficiently. To serve this market effectively, we need:

1. **Erasure-resistant storage** (ideally S3-compatible) located geographically close to users for optimal transfer speeds
2. **Direct mobile app ↔ S3 transfers** for bulk data, bypassing the Immich server to maximize throughput
3. **Flexible compute placement** where the Immich server can run on minimal infrastructure since it's not handling large file transfers

This architecture would address several pain points mentioned in this thread:
- **Cost optimization** (as @CrushedAsian255 and others noted) - pay for actual storage used vs. pre-provisioned disk space
- **Scalability** without infrastructure complexity 
- **Performance** through direct client-to-storage transfers for large files
- **Reliability** with proper object storage redundancy

The current workarounds using rclone/s3fs, while creative, introduce the performance and reliability issues many of you have documented. Native S3 support with direct mobile client integration would eliminate these bottlenecks entirely.

For Immich to become a truly compelling Google Photos alternative for the mainstream market, this architectural shift feels essential. I'm very interested in contributing to make this happen, particularly ensuring the implementation enables direct mobile-to-S3 transfers rather than proxying through the server.

~I noticed the recent discussion in #19979 where @danieldietzler mentioned that native S3 support is being actively worked on and you're hiring for it. This is fantastic news! My question there was specifically about whether the implementation will support direct client-to-S3 transfers for primary storage (not just backup). This architectural approach would be crucial for achieving the performance and cost benefits outlined above.~ 

Would the team be open to discussing the technical approach being considered? I'd be happy to contribute insights from the consulting perspective on what would make this most valuable for broader adoption.

**Reply by danieldietzler** [reply]
> I noticed the recent discussion in https://github.com/immich-app/immich/discussions/19979 where @danieldietzler mentioned that native S3 support is being actively worked on and you're hiring for it.

What? I said we're going to build out a one-click (hosted) backup solution. The Immich-side part of it will be compatible with multiple storage backends, including S3. That means you'll be able to back up Immich to a S3 location. Completely unrelated to storage backend S3 support.

**Reply by diraneyya** [reply]
Yes I realised that now, after reading your explanation over there. I apologise.

**Reply by sweepies** [reply]
Thanks for the writeup. The space as a whole is indeed moving away from locally provisioned storage. Even homelabbers are more often running object storage locally because it's easier to manage when using powerful tech like Kubernetes and declarative infrastructure.

Ente Photos chose this route from the beginning, and it happens to work very well. But I want to use *Immich* because I like its model of central compute for things like ML, more than I need end-to-end encryption.

Ente is open source -- maybe we can take inspiration?

**Comment by bo0tzz** [comment]
I'm going to close this in favour of #4445, as I don't think we'll ever be adding native support for any other backends.

**Reply by sweepies** [reply]
...why?

I, like others, am limited in my use of Immich without this feature. I LOVE Immich, so this makes me quite disappointed.

I run my apps in the cloud due to network constraints and geographically spread out users. I won't be able to properly utilize Immich unless I want to spend hundreds of dollars monthly on locally attached or block storage.

Why must we keep re-iterating this, only to be treated as second class citizens? Is my use case not intended/not supported by the Immich team?

While an S3 *backup* feature is nice, most of us already have backup solutions in place for our other things. For this to be prioritized as a "nice to have" over something that would allow more people to use Immich, I'm not sure I can understand.

As someone whom financially backed Immich for years, *please* reconsider.

**Reply by danieldietzler** [reply]
@sweepies what other storage backend would you care about besides S3 (and probably more importantly; why? And why not S3)?

**Reply by sweepies** [reply]
> @sweepies what other storage backend would you care about besides S3 (and probably more importantly; why? And why not S3)?

Apologies, I thought that comment linked back to the thread I came from - about S3 as a backup option only. I'm excited about S3 as a backend!
